keps/sig-storage/20191031-storage-capacity-constraints-for-pod-scheduling.md

```diff
@@ -0,0 +1,935 @@
+  nodes:
+  - node-2
+```
+
+##### Example: affect of storage classes
+
+```
+apiVersion: storage.k8s.io/v1alpha1
+kind: CSIStoragePool
+metadata:
+  name: sp-9c17f6fc-6ada-488f-9d44-c5d63ecdf7a9
+spec:
+  driverName: lvm
+status:
+  classes:
+  - capacity: 256G
+    storageClassName: striped
+  - capacity: 128G
+    storageClassName: mirrored
+  nodes:
+  - node-1
+```
+
+##### Example: network attached storage
+
+```
+apiVersion: storage.k8s.io/v1alpha1
+kind: CSIStoragePool
+metadata:
+  name: sp-b0963bb5-37cf-415d-9fb1-667499172320
+spec:
+  driverName: pd.csi.storage.gke.io
+status:
+  classes:
+  - capacity: 128G
+    storageClassName: <fallback>
+  nodeTopology:
+    nodeSelectorTerms:
+    - matchExpressions:
+      - key: failure-domain.beta.kubernetes.io/region
+        operator: In
+        values:
+        - us-east-1
+
+apiVersion: storage.k8s.io/v1alpha1
+kind: CSIStoragePool
+metadata:
+  name: sp-64103396-0d32-11ea-945c-e3ede5f0f3ae
+spec:
+  driverName: pd.csi.storage.gke.io
+status:
+  classes:
+  - capacity: 256G
+    storageClassName: <fallback>
+  nodeTopology:
+    nodeSelectorTerms:
+    - matchExpressions:
+      - key: failure-domain.beta.kubernetes.io/region
+        operator: In
+        values:
+        - us-west-1
+```
+
+#### CSIDriver.spec.storageCapacity
+
+A new field `storageCapacity` of type `boolean` with default `false`
+in
+[CSIDriver.spec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/#csidriverspec-v1beta1-storage-k8s-io)
+indicates whether a driver deployment will create `CSIStoragePool`
+objects. This allows the scheduler to determine whether it needs to
+consider available capacity for volumes of that driver or can proceed
+without it (as it does now).
+
+
+### Updating capacity information with external-provisioner
+
+Most (if not all) CSI drivers already get deployed on Kubernetes
+together with the external-provisioner which then handles volume
+provisioning via PVC. However, drivers which only handle inline
+ephemeral volumes currently don’t need that and thus get deployed
+without it.
+
+For the sake of avoiding yet another sidecar, the proposal is to
+extend the external-provisioner such that it can also be deployed
+alongside a CSI driver on each node and then just handles
+`CSIStoragePool` creation and updating. This leads to different mode
+of operations:
+
+#### Only local volumes
+
+This mode of operation is expected to be used by CSI drivers that need
+to track capacity per node and only support ephemeral inline volumes
+and/or persistent volumes with delayed binding (as proposed for the
+csi-driver-host-path in
+https://github.com/kubernetes-csi/external-provisioner/pull/367).
+
+The CSI driver has to implement the CSI controller service and its
+`GetCapacity` call. Its deployment has to add the external-provisioner
+to the daemon set and enable the per-node capacity tracking with
+`--enable-capacity=local-ephemeral`, `--enable-capacity=local-persistent`,
```

Comment:
I'm not sure we should have so many different modes of operation. This will make maintaining the provisioner very complex.

I think we should try to simplify things and require that the plugin supports Topology in order to use this feature, and only have two modes of operation "local" and "not local". This could align with the PR that's trying to allow running the external-provisioner as a daemonset: https://github.com/kubernetes-csi/external-provisioner/pull/367

Reply:
> I think we should try to simplify things and require that the plugin supports Topology

I'm fine with that. I added `global-*` for the sake of completeness, not necessarily because I think that it is important.

But I think we should keep the flags for ephemeral: without it, external-provisioner won't know whether calling `GetCapacity` without parameters and creation the special `<ephemeral>` entry really makes sense.

We can drop the "persistent" flag and instead assume that if storage classes are defined for the driver, those really need to be considered.

I'll do a revision without node list and less special cases. That's going to be less work to implement, too, and we can still extend it again later.

Reply:
I've revised the configuration of the external-provisioner, see 86cf87dc7a2e58fcfec9034180d14ef6bab29fcb.